README.md
# CenterTrack-HarDNet_JL
The CenterTrack detector with the HarDNet backbone for edge detection

This repo was forked from [CenterTrack](https://github.com/xingyizhou/CenterTrack). Changes have been made to incorporate the Harmonic Dense Net backbone used in [CenterNet-HarDNet](https://github.com/PingoLH/CenterNet-HarDNet). These changes are outlined in brief in the [change log](/change_log.md). The investigative process and steps used to determine them are described in detail in these [notes](/adding_hardnet_backbone_notes.md).


change_log.md
# Simple summary of changes made to the CenterTrack repository

This repository was forked from [CenterTrack](https://github.com/xingyizhou/CenterTrack) and modified to use the HarDNet backbone used in the [CenterNet-HarDNet](https://github.com/PingoLH/CenterNet-HarDNet) repository.

Details on how we arrived at the following changes are provided in these [notes](/adding_hardnet_backbone_notes.md).

In the following, `repo_root` refers to the root of this repository.

## Changes to the repository directory structure

1. Added a new directories `repo_root/models` and `repo_root/data` and included `.gitignore` files to not track anything inside
    - `models` is to store pre-trained `.pth` model files
    - `data` is to store symbolic links to the COCO2017 dataset

## Changes needed from updated modules and APIs

1. In `repo_root/src/lib/model/networks/backbones/mobilenet.py`:
    - Changed import of `load_state_dict_from_url`

2. In `repo_root/src/lib/model/utils`:
    - Changed import of `linear_assignment`


## Changes needed to incorporate the HarDNet backbone

1. In `repo_root/src/lib/model/model.py`:
    - Added import of `HarDNetSeg` and corresponding entry in `_network_factory`

2. In `repo_root/src/lib/model/networks`:
    - Copied file `hardnet.py` from the `CenterNet-HarDNet` repository located at `CenterNet-HarDNet/src/lib/models/networks/hardnet.py`
    - Fixed inconsistent indentation throughout file

3. In the new `repo_root/src/lib/model/networks/hardnet.py`:
    - Changed definition of `HarDNetSeg` class to accomodate its import in Step 1 above
    - Modified line in for loop defining the network head due to a variable name change from previous step
    - Added additional layers in `HarDNetSeg.__init__` for the input current and previous image, and previous heatmap
    - Added new lines at the beginning of `HarDNetSeg.forward` method to account for combining current and prior image information
    - Modified `for` loop in `HarDNetSeg.forward` to skip the first layer due to above necessary step in CenterTrack which modifies the first layer pass

4. In `repo_root/src`:
    - Copied pre-trained model files `hartnet68_base.pth` and `hartnet85_base.pth` from the `CenterNet-HarDNet` repository located at `CenterNet-HarDNet/src`

5. In `repo_root/models`:
    - Downloaded pretrained model `centernet_hardnet85_coco_608.pth` from the `CenterNet-HarDNet` repository

## Changes needed to train the network

1. In `repo_root/src/lin/logger.py`:
    - Added additional argument `--always` to subprocess call (this is not related to HarDNet, but fixed an error that arose in training)

## Training command:

So far, we have tested out the following training command, run inside `repo_root/src`:
```
$ python3 main.py tracking --exp_id coco_tracking --tracking --arch hardnet_85 --head_conv 256 --gpus 0 --batch_size 1 --lr 5e-4 --num_workers 16 --pre_hm --shift 0.05 --scale 0.05 --hm_disturb 0.05 --lost_disturb 0.4 --fp_disturb 0.1 --num_epochs 1
```

This is a modified version of the command found in the `CenterTrack/experiments/coco_tracking.sh` script that was used to train the DLA model we have been using in IRIS. We will test out other flags that can be set to generate validation metrics and save the model periodically or even after every epoch:
```
--save_point		# String of integers for epochs to save model at. Use this or:
--save_all		# Saves model to disk after each epoch
--eval_val		# True/False to evaluate model on validation set
--val_intervals x	# Epoch interval to run validation (when epoch is in save_point 
                        #   list, or epoch % val_intervals == 0)
```

For overfitting, it has been suggested by the developers of the CenterNet repository (on which CenterTrack was made by the same developers) to use validation AP over the typical loss, as the validation loss may not behave normally (decrease for a while and then increase, see [this issue](https://github.com/xingyizhou/CenterNet/issues/148) for example and more details)

## Changes needed to convert the model to TensorRT

Still in progress

src/demo_trt.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import _init_paths

import os
import cv2
from time import perf_counter

import torch
import numpy as np
from torch import nn
import torch.onnx

import tensorrt as trt
import pycuda.autoinit
import pycuda
import pycuda.driver as cuda

from opts import opts
from detectors.detector_factory import detector_factory
from pytorch_bn_fusion.bn_fusion import fuse_bn_recursively
from models.networks.hardnet import get_pose_net as get_hardnet
from models.model import load_model
from models.decode import ctdet_decode
from utils.debugger import Debugger
from utils.image import transform_preds

image_ext = ['jpg', 'jpeg', 'png', 'webp']
video_ext = ['mp4', 'mov', 'avi', 'mkv']
time_stats = ['tot', 'load', 'pre', 'net', 'dec', 'post', 'merge']

TRT_LOGGER = trt.Logger()  # This logger is required to build an engine


def get_img_np_nchw(filename):
    image = cv2.imread(filename)
    image_cv = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image_cv = cv2.resize(image_cv, (224, 224))
    miu = np.array([0.485, 0.456, 0.406])
    std = np.array([0.229, 0.224, 0.225])
    img_np = np.array(image_cv, dtype=float) / 255.
    r = (img_np[:, :, 0] - miu[0]) / std[0]
    g = (img_np[:, :, 1] - miu[1]) / std[1]
    b = (img_np[:, :, 2] - miu[2]) / std[2]
    img_np_t = np.array([r, g, b])
    img_np_nchw = np.expand_dims(img_np_t, axis=0)
    return img_np_nchw

class HostDeviceMem(object):
    def __init__(self, host_mem, device_mem):
        """Within this context, host_mom means the cpu memory and device means the GPU memory
        """
        self.host = host_mem
        self.device = device_mem

    def __str__(self):
        return "Host:\n" + str(self.host) + "\nDevice:\n" + str(self.device)

    def __repr__(self):
        return self.__str__()


def allocate_buffers(engine):
    inputs = []
    outputs = []
    bindings = []
    stream = cuda.Stream()
    for binding in engine:
        size = trt.volume(engine.get_binding_shape(binding)) * engine.max_batch_size
        dtype = trt.nptype(engine.get_binding_dtype(binding))
        # Allocate host and device buffers
        host_mem = cuda.pagelocked_empty(size, dtype)
        device_mem = cuda.mem_alloc(host_mem.nbytes)
        # Append the device buffer to device bindings.
        bindings.append(int(device_mem))
        # Append to the appropriate list.
        if engine.binding_is_input(binding):
            inputs.append(HostDeviceMem(host_mem, device_mem))
        else:
            outputs.append(HostDeviceMem(host_mem, device_mem))
    return inputs, outputs, bindings, stream


def get_engine(max_batch_size=1, onnx_file_path="", engine_file_path="", \
               fp16_mode=False, int8_mode=False, overwrite=False
               ):

    def build_engine(max_batch_size):

        explicit_batch = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
        
        with trt.Builder(TRT_LOGGER) as builder, \
                builder.create_network(explicit_batch) as network, \
                trt.OnnxParser(network, TRT_LOGGER) as parser:
            # NEW two lines: (builder.max_workspace_size is removed in tensorrt 8)
            builder_config = builder.create_builder_config()
            builder_config.max_workspace_size = 1 << 30
            # builder.max_workspace_size = 1 << 30  # Your workspace size
            
            # NEW lines to set precision
            builder.max_batch_size = max_batch_size
            if fp16_mode:
                builder_config.set_flag(trt.BuilderFlag.FP16)
            if int8_mode:
                builder_config.set_flag(trt.BuilderFlag.INT8)

            # builder.fp16_mode = fp16_mode  # Default: False
            # builder.int8_mode = int8_mode  # Default: False
            
            # if int8_mode:
            #     raise NotImplementedError

            # Parse model file
            if not os.path.exists(onnx_file_path):
                quit('ONNX file {} not found'.format(onnx_file_path))

            print('Loading ONNX file from path {}...'.format(onnx_file_path))
            with open(onnx_file_path, 'rb') as model:
                print('Beginning ONNX file parsing')
                res = parser.parse(model.read())
            if res:
              print('Completed parsing of ONNX file')
              print('# Layers = ', network.num_layers)
              print('Building an engine from file {}; this may take a while...'.format(onnx_file_path))
            else:
              print('Parse Failed, Layers = ', network.num_layers)
              exit()              

            # NEW lines to build engine (builder.build_cuda_engine() was removed in tensorrt 8)
            plan = builder.build_serialized_network(network, builder_config)
            with trt.Runtime(TRT_LOGGER) as runtime:
              engine = runtime.deserialize_cuda_engine(plan)
            # engine = builder.build_cuda_engine(network)
            print("Completed creating Engine")

            if engine_file_path:
                with open(engine_file_path, "wb") as f:
                    f.write(engine.serialize())
            return engine

    if os.path.exists(engine_file_path) and not overwrite:
        print("Reading engine from file {}".format(engine_file_path))
        with open(engine_file_path, "rb") as f, trt.Runtime(TRT_LOGGER) as runtime:
            return runtime.deserialize_cuda_engine(f.read())
    elif not os.path.exists(onnx_file_path):
        print('Cannot find any ONNX file or TRT file')
        exit()
    else:
        return build_engine(max_batch_size)


def do_inference(context, bindings, inputs, outputs, stream, batch_size=1):
    # Transfer data from CPU to the GPU.
    [cuda.memcpy_htod_async(inp.device, inp.host, stream) for inp in inputs]
    # Run inference.
    context.execute_async(batch_size=batch_size, bindings=bindings, stream_handle=stream.handle)
    # Transfer predictions back from the GPU.
    [cuda.memcpy_dtoh_async(out.host, out.device, stream) for out in outputs]
    # Synchronize the stream
    stream.synchronize()
    # Return only the host outputs.
    return [out.host for out in outputs]


def postprocess_the_outputs(h_outputs, shape_of_output):
    h_outputs = h_outputs.reshape(*shape_of_output)
    return h_outputs

class CenterHarDNet(nn.Module):
  def __init__(self, num_layers, opt):
    super().__init__()
    heads = {'hm': 80,'wh': 4}
    model = get_hardnet(num_layers=num_layers, heads=opt.heads, head_conv=opt.head_conv, trt=True)
    if opt.load_model:
      model = load_model(model, opt.load_model)
    model = fuse_bn_recursively(model)
    # model.v2_transform()
    self.model = model
    mean = np.array(opt.mean, dtype=np.float32).reshape( 1, 3, 1, 1)
    std = np.array(opt.std, dtype=np.float32).reshape( 1, 3, 1, 1)
    self.mean = nn.Parameter(torch.from_numpy(mean))
    self.std = nn.Parameter(torch.from_numpy(std))
    self.max_per_image = 100
    self.num_classes = opt.num_classes
    self.scales = opt.test_scales
    self.opt = opt
    
  def forward(self, x):
    x = ( x/(torch.ones(x.shape, dtype=x.dtype, device=x.device)*255) - self.mean) / self.std
    out = self.model(x)

    hm = torch.sigmoid(out[0])
    wh = out[1]
    reg = wh[:,2:,:,:]
    wh  = wh[:,:2,:,:]
    dets = ctdet_decode(hm, wh, reg=reg, cat_spec_wh=self.opt.cat_spec_wh, K=self.opt.K, trt=True)
    
    return dets

    
def show_det(dets, image, det_size, debugger, opt, pause=False, name=None):
  dets = dets.reshape(1, -1, dets.shape[2])
  h,w = image.shape[0:2]
  debugger.add_img(image, img_id='ctdet')
  
  c = np.array([w / 2, h / 2], dtype=np.float32)
  s = np.array([w, h], dtype=np.float32)
  dets[0, :,  :2] = transform_preds( dets[0, :, 0:2], c, s, det_size)
  dets[0, :, 2:4] = transform_preds( dets[0, :, 2:4], c, s, det_size)
  classes = dets[0, :, -1]
  
  for j in range(opt.num_classes):
    inds = (classes == j)
    top_preds = dets[0, inds, :5].astype(np.float32)
    for bbox in top_preds:
      if bbox[4] > opt.vis_thresh:
        debugger.add_coco_bbox(bbox[:4], j , bbox[4], img_id='ctdet')
        
  if name:
    # print('detecting:', name)
    debugger.save_all_imgs( path='./', prefix=name)
  else:
    debugger.show_all_imgs(pause=pause)


def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

def demo(opt):
  os.environ['CUDA_VISIBLE_DEVICES'] = opt.gpus_str
  opt.debug = max(opt.debug, 1)
  debugger = Debugger(dataset=opt.dataset, ipynb=(opt.debug==3),
                        theme=opt.debugger_theme)

  model = CenterHarDNet(85, opt).cuda()
  model.eval()
  
  image_size = (opt.input_w, opt.input_h)
  det_size = [image_size[0]//opt.down_ratio, image_size[1]//opt.down_ratio]
  
  onnx_model_path = "ctdet_%s_%dx%d.onnx"%(opt.arch,image_size[0], image_size[1])
  trt_engine_path = "ctdet_%s_%dx%d.trt"%(opt.arch,image_size[0], image_size[1])
  
  x = torch.randn((1, 3, image_size[1], image_size[0])).float().cuda()
  
  if opt.load_trt:
    trt_engine_path = opt.load_trt
    engine = get_engine(1, "", trt_engine_path, fp16_mode=True)
  else:
    if not opt.load_model:
      print('Please load model with --load_model')
      exit()
    print('\nStep 1: Converting ONNX... (PyTorch>1.3 is required)')
    torch.onnx.export(model,               # model being run
                      x,                         # model input (or a tuple for multiple inputs)
                      onnx_model_path,           # where to save the model (can be a file or file-like object)
                      export_params=True,        # store the trained parameter weights inside the model file
                      opset_version=11,          # the ONNX version to export the model to
                      do_constant_folding=True,  # whether to execute constant folding for optimization
                      input_names = ['input'],   # the model's input names
                      output_names = ['output'], # the model's output names
                      dynamic_axes=None)
    print('\nStep 2: Converting TensorRT... ')
    engine = get_engine(1, onnx_model_path, trt_engine_path, fp16_mode=True, int8_mode=False, overwrite=True)
    
  outs = model(x)
  out_shape = outs.shape
  
  context = engine.create_execution_context()
  inputs, outputs, bindings, stream = allocate_buffers(engine)
  
  if not opt.demo:
    exit() 
                                
  if opt.demo == 'webcam' or \
    opt.demo[opt.demo.rfind('.') + 1:].lower() in video_ext:
    cam = cv2.VideoCapture(0 if opt.demo == 'webcam' else opt.demo)
    cam.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
    cam.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
    cam.set(cv2.CAP_PROP_FPS, 30)
    
    while True:
        t = perf_counter()
        _, img = cam.read()
        img_x = cv2.resize(img, image_size)
        x = img_x.transpose(2, 0, 1).reshape(1, 3, image_size[1], image_size[0]).astype(np.float32)
        inputs[0].host = x.reshape(-1)

        t2 = perf_counter()
        outs = do_inference(context, bindings=bindings, inputs=inputs, outputs=outputs, stream=stream)
        dets = outs[0].reshape(out_shape)
        t3 = perf_counter()

        show_det(dets, img, det_size, debugger, opt, False)

        print(' Latency = %.2f ms  (net=%.2f ms)'%((perf_counter()-t)*1000, (t3-t2)*1000))
        if cv2.waitKey(1) == 27:
            return  # esc to quit
  else:
    if os.path.isdir(opt.demo):
      image_names = []
      ls = os.listdir(opt.demo)
      for file_name in sorted(ls):
          ext = file_name[file_name.rfind('.') + 1:].lower()
          if ext in image_ext:
              image_names.append(os.path.join(opt.demo, file_name))
    else:
      image_names = [opt.demo]
    
    for (image_name) in image_names:

      t_start = perf_counter()
      img = cv2.imread(image_name)
      img_x = cv2.resize(img, image_size)

      x = img_x.transpose(2, 0, 1).reshape(1, 3, image_size[1], image_size[0]).astype(np.float32)
      inputs[0].host = x.reshape(-1)

      t_inf_start = perf_counter()
      outs = do_inference(context, bindings=bindings, inputs=inputs, outputs=outputs, stream=stream)
      t_inf = perf_counter()-t_inf_start
      # print(outs)
      dets = outs[0].reshape(out_shape)
      # print(dets)
      show_det(dets, img, det_size, debugger, opt, True, os.path.basename(image_name))
      t_tot = perf_counter()-t_start

      print(f' Image: {os.path.basename(image_name)} | Timings: | tot {t_tot : .4f} | net {t_inf : .4f}')
      
if __name__ == '__main__':
  opt = opts().init()
  demo(opt)


src/lib/logger.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

# Code referenced from https://gist.github.com/gyglim/1f8dfb1b5c82627ae3efcfbbadb9f514
import os
import time
import sys
import torch
import subprocess
USE_TENSORBOARD = True
try:
  import tensorboardX
  print('Using tensorboardX')
except:
  USE_TENSORBOARD = False

class Logger(object):
  def __init__(self, opt):
    """Create a summary writer logging to log_dir."""
    if not os.path.exists(opt.save_dir):
      os.makedirs(opt.save_dir)
    if not os.path.exists(opt.debug_dir):
      os.makedirs(opt.debug_dir)
   
    time_str = time.strftime('%Y-%m-%d-%H-%M')

    args = dict((name, getattr(opt, name)) for name in dir(opt)
                if not name.startswith('_'))
    file_name = os.path.join(opt.save_dir, 'opt.txt')
    with open(file_name, 'wt') as opt_file:
      opt_file.write('==> commit hash: {}\n'.format(
        # subprocess.check_output(["git", "describe"])))  # OLD line
        subprocess.check_output(["git", "describe", "--always"]))) # NEW line
      opt_file.write('==> torch version: {}\n'.format(torch.__version__))
      opt_file.write('==> cudnn version: {}\n'.format(
        torch.backends.cudnn.version()))
      opt_file.write('==> Cmd:\n')
      opt_file.write(str(sys.argv))
      opt_file.write('\n==> Opt:\n')
      for k, v in sorted(args.items()):
        opt_file.write('  %s: %s\n' % (str(k), str(v)))
          
    log_dir = opt.save_dir + '/logs_{}'.format(time_str)
    if USE_TENSORBOARD:
      self.writer = tensorboardX.SummaryWriter(log_dir=log_dir)
    else:
      if not os.path.exists(os.path.dirname(log_dir)):
        os.mkdir(os.path.dirname(log_dir))
      if not os.path.exists(log_dir):
        os.mkdir(log_dir)
    self.log = open(log_dir + '/log.txt', 'w')
    try:
      os.system('cp {}/opt.txt {}/'.format(opt.save_dir, log_dir))
    except:
      pass
    self.start_line = True

  def write(self, txt):
    if self.start_line:
      time_str = time.strftime('%Y-%m-%d-%H-%M')
      self.log.write('{}: {}'.format(time_str, txt))
    else:
      self.log.write(txt)  
    self.start_line = False
    if '\n' in txt:
      self.start_line = True
      self.log.flush()
  
  def close(self):
    self.log.close()
  
  def scalar_summary(self, tag, value, step):
    """Log a scalar variable."""
    if USE_TENSORBOARD:
      self.writer.add_scalar(tag, value, step)


src/lib/model/model.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import torchvision.models as models
import torch
import torch.nn as nn
import os

from .networks.dla import DLASeg
from .networks.resdcn import PoseResDCN
from .networks.resnet import PoseResNet
from .networks.dlav0 import DLASegv0
from .networks.generic_network import GenericNetwork
from .networks.hardnet import HarDNetSeg # NEW line

_network_factory = {
  'hardnet': HarDNetSeg, # NEW line
  'resdcn': PoseResDCN,
  'dla': DLASeg,
  'res': PoseResNet,
  'dlav0': DLASegv0,
  'generic': GenericNetwork
}

def create_model(arch, head, head_conv, opt=None):
  num_layers = int(arch[arch.find('_') + 1:]) if '_' in arch else 0
  arch = arch[:arch.find('_')] if '_' in arch else arch
  model_class = _network_factory[arch]
  model = model_class(num_layers, heads=head, head_convs=head_conv, opt=opt)
  return model

def load_model(model, model_path, opt, optimizer=None):
  start_epoch = 0
  checkpoint = torch.load(model_path, map_location=lambda storage, loc: storage)
  print('loaded {}, epoch {}'.format(model_path, checkpoint['epoch']))
  state_dict_ = checkpoint['state_dict']
  state_dict = {}
   
  # convert data_parallal to model
  for k in state_dict_:
    if k.startswith('module') and not k.startswith('module_list'):
      state_dict[k[7:]] = state_dict_[k]
    else:
      state_dict[k] = state_dict_[k]
  model_state_dict = model.state_dict()

  # check loaded parameters and created model parameters
  for k in state_dict:
    if k in model_state_dict:
      if (state_dict[k].shape != model_state_dict[k].shape) or \
        (opt.reset_hm and k.startswith('hm') and (state_dict[k].shape[0] in [80, 1])):
        if opt.reuse_hm:
          print('Reusing parameter {}, required shape{}, '\
                'loaded shape{}.'.format(
            k, model_state_dict[k].shape, state_dict[k].shape))
          if state_dict[k].shape[0] < state_dict[k].shape[0]:
            model_state_dict[k][:state_dict[k].shape[0]] = state_dict[k]
          else:
            model_state_dict[k] = state_dict[k][:model_state_dict[k].shape[0]]
          state_dict[k] = model_state_dict[k]
        else:
          print('Skip loading parameter {}, required shape{}, '\
                'loaded shape{}.'.format(
            k, model_state_dict[k].shape, state_dict[k].shape))
          state_dict[k] = model_state_dict[k]
    else:
      print('Drop parameter {}.'.format(k))
  for k in model_state_dict:
    if not (k in state_dict):
      print('No param {}.'.format(k))
      state_dict[k] = model_state_dict[k]
  model.load_state_dict(state_dict, strict=False)

  # resume optimizer parameters
  if optimizer is not None and opt.resume:
    if 'optimizer' in checkpoint:
      # optimizer.load_state_dict(checkpoint['optimizer'])
      start_epoch = checkpoint['epoch']
      start_lr = opt.lr
      for step in opt.lr_step:
        if start_epoch >= step:
          start_lr *= 0.1
      for param_group in optimizer.param_groups:
        param_group['lr'] = start_lr
      print('Resumed optimizer with start lr', start_lr)
    else:
      print('No optimizer parameters in checkpoint.')
  if optimizer is not None:
    return model, optimizer, start_epoch
  else:
    return model

def save_model(path, epoch, model, optimizer=None):
  if isinstance(model, torch.nn.DataParallel):
    state_dict = model.module.state_dict()
  else:
    state_dict = model.state_dict()
  data = {'epoch': epoch,
          'state_dict': state_dict}
  if not (optimizer is None):
    data['optimizer'] = optimizer.state_dict()
  torch.save(data, path)



src/lib/utils/tracker.py
import numpy as np
# from sklearn.utils.linear_assignment_ import linear_assignment # OLD import
from scipy.optimize import linear_sum_assignment as linear_assignment # NEW line
from numba import jit
import copy

class Tracker(object):
  def __init__(self, opt):
    self.opt = opt
    self.reset()

  def init_track(self, results):
    for item in results:
      if item['score'] > self.opt.new_thresh:
        self.id_count += 1
        # active and age are never used in the paper
        item['active'] = 1
        item['age'] = 1
        item['tracking_id'] = self.id_count
        if not ('ct' in item):
          bbox = item['bbox']
          item['ct'] = [(bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2]
        self.tracks.append(item)

  def reset(self):
    self.id_count = 0
    self.tracks = []

  def step(self, results, public_det=None):
    N = len(results)
    M = len(self.tracks)

    dets = np.array(
      [det['ct'] + det['tracking'] for det in results], np.float32) # N x 2
    track_size = np.array([((track['bbox'][2] - track['bbox'][0]) * \
      (track['bbox'][3] - track['bbox'][1])) \
      for track in self.tracks], np.float32) # M
    track_cat = np.array([track['class'] for track in self.tracks], np.int32) # M
    item_size = np.array([((item['bbox'][2] - item['bbox'][0]) * \
      (item['bbox'][3] - item['bbox'][1])) \
      for item in results], np.float32) # N
    item_cat = np.array([item['class'] for item in results], np.int32) # N
    tracks = np.array(
      [pre_det['ct'] for pre_det in self.tracks], np.float32) # M x 2
    dist = (((tracks.reshape(1, -1, 2) - \
              dets.reshape(-1, 1, 2)) ** 2).sum(axis=2)) # N x M

    invalid = ((dist > track_size.reshape(1, M)) + \
      (dist > item_size.reshape(N, 1)) + \
      (item_cat.reshape(N, 1) != track_cat.reshape(1, M))) > 0
    dist = dist + invalid * 1e18
    
    if self.opt.hungarian:
      item_score = np.array([item['score'] for item in results], np.float32) # N
      dist[dist > 1e18] = 1e18
      matched_indices = linear_assignment(dist)
    else:
      matched_indices = greedy_assignment(copy.deepcopy(dist))
    unmatched_dets = [d for d in range(dets.shape[0]) \
      if not (d in matched_indices[:, 0])]
    unmatched_tracks = [d for d in range(tracks.shape[0]) \
      if not (d in matched_indices[:, 1])]
    
    if self.opt.hungarian:
      matches = []
      for m in matched_indices:
        if dist[m[0], m[1]] > 1e16:
          unmatched_dets.append(m[0])
          unmatched_tracks.append(m[1])
        else:
          matches.append(m)
      matches = np.array(matches).reshape(-1, 2)
    else:
      matches = matched_indices

    ret = []
    for m in matches:
      track = results[m[0]]
      track['tracking_id'] = self.tracks[m[1]]['tracking_id']
      track['age'] = 1
      track['active'] = self.tracks[m[1]]['active'] + 1
      ret.append(track)

    if self.opt.public_det and len(unmatched_dets) > 0:
      # Public detection: only create tracks from provided detections
      pub_dets = np.array([d['ct'] for d in public_det], np.float32)
      dist3 = ((dets.reshape(-1, 1, 2) - pub_dets.reshape(1, -1, 2)) ** 2).sum(
        axis=2)
      matched_dets = [d for d in range(dets.shape[0]) \
        if not (d in unmatched_dets)]
      dist3[matched_dets] = 1e18
      for j in range(len(pub_dets)):
        i = dist3[:, j].argmin()
        if dist3[i, j] < item_size[i]:
          dist3[i, :] = 1e18
          track = results[i]
          if track['score'] > self.opt.new_thresh:
            self.id_count += 1
            track['tracking_id'] = self.id_count
            track['age'] = 1
            track['active'] = 1
            ret.append(track)
    else:
      # Private detection: create tracks for all un-matched detections
      for i in unmatched_dets:
        track = results[i]
        if track['score'] > self.opt.new_thresh:
          self.id_count += 1
          track['tracking_id'] = self.id_count
          track['age'] = 1
          track['active'] =  1
          ret.append(track)
    
    for i in unmatched_tracks:
      track = self.tracks[i]
      if track['age'] < self.opt.max_age:
        track['age'] += 1
        track['active'] = 0
        bbox = track['bbox']
        ct = track['ct']
        v = [0, 0]
        track['bbox'] = [
          bbox[0] + v[0], bbox[1] + v[1],
          bbox[2] + v[0], bbox[3] + v[1]]
        track['ct'] = [ct[0] + v[0], ct[1] + v[1]]
        ret.append(track)
    self.tracks = ret
    return ret

def greedy_assignment(dist):
  matched_indices = []
  if dist.shape[1] == 0:
    return np.array(matched_indices, np.int32).reshape(-1, 2)
  for i in range(dist.shape[0]):
    j = dist[i].argmin()
    if dist[i][j] < 1e16:
      dist[:, j] = 1e18
      matched_indices.append([i, j])
  return np.array(matched_indices, np.int32).reshape(-1, 2)


src/centertrack-hardnet_sandbox.py
import _init_paths

import os
import sys
import torch
import cv2

from detector import Detector
from opts import opts

# Function called by camnetGPU and camnetCPU initilizations
def load_detector(centertrackdir='/home/jugaadlabs/JL/centertrack_testing/CenterTrack', img_width=720, img_height=616):
    args = [#'task=tracking',
           #'--arch=hardnet_85',
           #'--load_model={}'.format(os.path.join(centertrackdir, 'models', 'coco_tracking.pth')),
           '--input_w={}'.format(img_width),
           '--input_h={}'.format(img_height)]

    opt = opts().init(args)
    
    os.environ['CUDA_VISIBLE_DEVICES'] = opt.gpus_str

    # print(f"opt.task: {opt.task}")
    detector = Detector(opt)

    return detector

# Function called by camnetGPU.runTarget
def run_detector_gpu(detector, image):
	"""
	Function to perform object detection on the GPU
	"""

	with torch.no_grad():
		ret = detector.run(image)

	# Return the 'results' value, which is a dictionary of the form 
	# {'score': x.xx, 
	#  'class': int, 
	#  'ct': array([xx, yy], dtype=float32), 
	#  'tracking': array([xx, yy], dtype=float32), 
	#  'bbox': array([x1, y1, x2, y2], dtype=float32), 
	#  'tracking_id': int, 
	#  'age': int (likely 1), 
	#  'active': int}
	return ret['results']


if __name__ == "__main__":
    sys.argv.append('task=tracking')
    # sys.argv.append('--input_w=720')
    # sys.argv.append('--input_h=616')
    # sys.argv.append('--input_w=704')
    # sys.argv.append('--input_h=608')

    # Un-comment for HarDNet-68:
    # sys.argv.append('--arch=hardnet_68')
    # Un-comment for HarDNet-85:
    sys.argv.append('--arch=hardnet_85')
    sys.argv.append('--head_conv=256')
    sys.argv.append('--load_model={}'.format(os.path.join('/home/jugaadlabs/JL/CenterTrack-HarDNet_JL', 'models', 'centernet_hardnet85_coco_608.pth')))
    


    # Un-comment for DLA-34:
    # sys.argv.append('--arch=dla_34')
    # sys.argv.append('--load_model={}'.format(os.path.join('/home/jugaadlabs/JL/centertrack_testing/CenterTrack', 'models', 'coco_tracking.pth')))



    opt = opts().init()
    
    os.environ['CUDA_VISIBLE_DEVICES'] = opt.gpus_str
    print('===== opt properties =====')
    print(f"opt.gpus_str: {opt.gpus_str}")
    print(f"opt.arch: {opt.arch}")
    print(f"opt.task: {opt.task}")
    print(f"opt.heads: {opt.heads}")
    print(f"opt.head_conv: {opt.head_conv}")
    print(f"opt.input_h: {opt.input_h}")
    print(f"opt.input_w: {opt.input_w}")
    print('==========================')
    detector = Detector(opt)



    # sys.argv.append('--load_model={}'.format(os.path.join('/home/jugaadlabs/JL/centertrack_testing/CenterTrack', 'models', 'coco_tracking.pth')))
    images_path = '/home/jugaadlabs/JL/ML_benchmarking/benchmarking_images/parking_lot_driving/driveby_pass'
    image_filenames = [f'driveby_left_wide_frame_{x}.png' for x in range(1326, 1426)]

    image = cv2.imread(os.path.join(images_path, image_filenames[50]))
    # cv2.imshow('test', image)
    # cv2.waitKey()
    # cv2.destroyWindow('test')
    # print(f'image.shape: {image.shape}')
    # resized_image = cv2.resize(image, (704, 608))


    # # detector = load_detector(arch='dla_34', model='ctdet_coco_dla_1x.pth', img_width=704, img_height=608)
    # # detector = load_detector(arch='hourglass', model='ctdet_coco_hg.pth', img_width=704, img_height=608)
    # detector = load_detector()

    results = detector.run(image)

    if len(results['results']) > 0:
      for res in results['results']:
          print(res)
    else:
      print(results)